### PR TITLE
Make borsh-py a real thing

### LINKED ISSUE #1: Implement borsh serialize feature

### LINKED ISSUE #2: Implement borsh deserialize feature

### LINKED ISSUE #3: Is it alive project?
I've installed it via pypi, but it's just an empty __init__.py file.

### PR DESCRIPTION
**Pull request by kevinheavey**
Good news everyone: borsh-py is here. This PR overwrites everything in the repo - I started from scratch but I think this makes sense as the home for borsh-py on GitHub, hence the PR rather than a new project. I'm hoping you'll add me as a maintainer to the repo.

What I've done in one sentence: 

Implemented the borsh spec per https://borsh.io/, using the very powerful [construct](https://construct.readthedocs.io/en/latest/) library.

Closes #1, closes #2, closes #3

### CODE REVIEW

**Comment by kevinheavey** [comment]
Also I should mention I still have to add CI for docs and for publishing to PyPI. That'll come in the next PR

**Review by frol — CHANGES_REQUESTED** [review]
@kevinheavey Wow, that is a great amount of work! :rocket: 

It is great to see a feature-complete borsh implementation in Python, but I have to admit, that I believe that it is not going to be intuitive for Python developers. I am happy to invite you as a maintainer once we sort out the design concerns and merge this PR :+1:

File: docs/index.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+## Usage
+
+Since `borsh-py` is built with `construct`, it serializes objects using `.build` and deserializes them using `.parse`. For example:
+
+```python
+>>> from borsh import U8, String, CStruct
+>>> animal = CStruct(
+...     "name" / String,
+...     "legs" / U8
+... )
```
[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/index.md:32]: I have indeed! Unfortunately the tooling isn't really there yet to make this happen.

Now, there is a [construct_typed](https://github.com/timrid/construct-typing) library that someone else built that adds dataclass, but it has a big health warning on it: `!!! EXPERIMENTAL VERSION !!!`. So I wouldn't want to bring that in for now. It's also not very convenient and doesn't make much sense unless you already know your way around `construct`. The animal example from above would end up looking something like this:

```python
import dataclasses
from borsh import DataclassMixin, DataclassStruct, csfield, String, U8

@dataclasses.dataclass
class Animal(DataclassMixin):
    name: str= csfield(String)
    legs: int= csfield(U8)

format = DataclassStruct(Image)
obj = Animal(
    name="Ferris",
    legs=6
)
print(format.build(obj))
```

So aside from the `!!! EXPERIMENTAL VERSION !!!` warning this does add a fair bit more code and confusion for users. But the "exoerimental" status is enough reason in itself not to use that library.

File: docs/types.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+## Dynamic sized arrays
+
+Dynamic arrays are implemented using the `Vec` function:
+
+```python
+>>> from borsh import Vec, U8
+>>> Vec(U8).build([1, 2, 3])
+b'\x03\x00\x00\x00\x01\x02\x03'
+>>> Vec(U8).parse(b'\x03\x00\x00\x00\x01\x02\x03')
+ListContainer([1, 2, 3])
```
[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/types.md:57]: ListContainer is just a list with some extra stuff to make `construct` work. So you can use it just like a list. I think unexpected things would break if we hacked the return type to be just a list

**Comment by kevinheavey** [comment]
Hey @frol thanks for the feedback! I'm adding individual responses to your comments, but the overarching theme is that the API design is based on the [construct](https://construct.readthedocs.io/en/latest/) library, because that's what borsh-py uses under the hood.

In my experience `construct` is by far the most popular library for doing this kind of work in Python, and while `construct`'s API is kinda weird it does work really well and is battle-tested. There would be an order of magnitude more work involved if we deviated from the `construct` approach, and if I were a prospective user who had prior experience with this kinda stuff I think I'd be more wary of something that didn't use `construct`.

In particular, `construct` has really good functionality for writing your own custom types, and someone who wants to extend borsh-py to serialize say, IP addresses, can drop down into `construct` land to do so. Come to think of it, I'll add an example to the docs showing how to do that.

[The dataclasses point is addressed in a reply above].

To summarise, I agree that the `construct` style is a little weird. But I also strongly believe the outcomes would be worse for everyone if we ditched it. And as someone who needs to use borsh-py, I doubt I would end up using a borsh-py that deviated from the `construct` approach. We don't know each other but please trust me as someone who has gone deep down this rabbit hole 🙏

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+>>> animal.build({"name": "Ferris", "legs": 6})
+b'\x06\x00\x00\x00Ferris\x06'
+>>> animal.parse(b'\x06\x00\x00\x00Ferris\x06')
+Container(name=u'Ferris', legs=6)
```
[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/index.md:36]: Worth mentioning that I think ctypes is a confusing thing to introduce for users here, plus it doesn't have all the types we need, plus it would require rewriting a lot of stuff

File: docs/types.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+## Fixed sized arrays
+
+`construct` gives us a nice `[]` syntax to represent fixed sized arrays. For example, an array of 3 u8 integers:
+
+```python
+>>> from borsh import U8
+>>> U8[3].build([1, 2, 3])
+b'\x01\x02\x03'
+>>> U8[3].parse(b'\x01\x02\x03')
+ListContainer([1, 2, 3])
```
[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/types.md:37]: Construct will actually accept array.array or tuple here. But I don't think changing the return type is a good idea because it will break some `construct` stuff.

It's very rare for people to use the `array` library in Python so I don't think it's worth even mentioning that one to users.

File: docs/types.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+>>> message.parse(b'\x01\x01\x00\x00\x00\x03\x00\x00\x00')
+EnumDef.Move(x=1, y=3)
+>>> message.build(message.enum.Write(("hello",)))
+b'\x02\x05\x00\x00\x00hello'
+>>> message.parse(b'\x02\x05\x00\x00\x00hello')
+EnumDef.Write(tuple_data=ListContainer(['hello']))
+>>> message.build(message.enum.ChangeColor((1, 2, 3)))
+b'\x03\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00'
+>>> message.parse(b'\x03\x01\x00\x00\x00\x02\x00\x00\x00\x03\x00\x00\x00')
+EnumDef.ChangeColor(tuple_data=ListContainer([1, 2, 3]))
```
[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/types.md:141]: Having a think about this (the decorator stuff won't happen though)

[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/types.md]: Added an `enum_name` parameter so we don't have that unpleasant `EnumDef` anymore:

```python
>>> message = Enum(
...     "Quit",
...     "Move" / CStruct("x" / I32, "y" / I32),
...     "Write" / TupleStruct(String),
...     "ChangeColor" / TupleStruct(I32, I32, I32),
...     enum_name="Message",
... )
>>> message.build(message.enum.Quit())
b'\x00'
>>> message.parse(b'\x00')
Message.Quit()

```

Making real runtime union types in Python is a bit involved, so I'm pretty content to just use what the `sumtypes` library gives us.

I think there's possibly a lot of work in making more elegant sum types in Python. However in the meantime I need borsh-py for stuff, which is why I just made something that resembles Rust enums where the variants are either unit struct, tuple struct or C-like struct.

If you want we could put a health warning in the Enum docs like `THIS API IS LIKELY TO CHANGE`. That'd work for me.

Btw this enum type comes with its own support for pattern matching: more in the sumtypes docs here https://sumtypes.readthedocs.io/en/latest/

[the remark this answers is not in the document]
Comment (kevinheavey) [reply docs/types.md]: Pattern matching example (not my design, just comes with sumtypes):

```python

from borsh import Enum, CStruct, TupleStruct, String, I32
from sumtypes import match

message = Enum(
    "Quit",
    "Move" / CStruct("x" / I32, "y" / I32),
    "Write" / TupleStruct(String),
    "ChangeColor" / TupleStruct(I32, I32, I32),
    enum_name="Message",
)


@match(message.enum)
class get_some_values(object):
    def Quit():
        return "quit"

    def Move(x, y):
        return y

    def Write(x):
        return x

    def ChangeColor(data):
        return data[2]


assert get_some_values(message.enum.Quit()) == "quit"
assert get_some_values(message.enum.Move(1, 2)) == 2
assert get_some_values(message.enum.Write(11)) == 11
assert get_some_values(message.enum.ChangeColor((1, 2, 3))) == 3

```

**Comment by kevinheavey** [comment]
Ok I've finished my batch of changes

**Comment by frol** [comment]
@kevinheavey Hey, I am sorry for the long delay, I am quite at capacity :pray: 

How do you feel if we name this package as `construct-borsh` or `borsh-construct` (is there a convention for construct-based implementations? are there any packages like this one out there? I would love to learn more about those)

**Comment by kevinheavey** [comment]
@frol sure let's call it `borsh-construct` then. Probably most convenient if we just rename this repository rather than creating a new one? It was empty before so it won't affect anyone but me.

If you change the repository name I'll add a commit to change what needs changing in the code

**Comment by SkidanovAlex** [comment]
Regarding names structs, what do people think about an approach like

```python
@borshify
class Animal:
    def __init__(name: String, breed: String, age: U32):
        pass
```

Where `@borshify` extracts the schema of the struct from the `__init__` method arguments?

**Comment by kevinheavey** [comment]
I think that would belong in another library tbh. At this point it’s quite clear what `borsh-construct` offers

**Comment by frol** [comment]
> sure let's call it borsh-construct then

Done. The repo is now renamed to `borsh-construct-py`. Let's update all the README and other references and get it merged :+1:

**Comment by kevinheavey** [comment]
Changes made 👍

**Review by frol — APPROVED** [review]
